.golangci.yml

```diff
@@ -41,6 +41,8 @@ linters-settings:
     forbid:
       - '^fmt\.Errorf(# use errors\.Errorf instead)?$'
       - '^logrus\.(Trace|Debug|Info|Warn|Warning|Error|Fatal)(f|ln)?(# use bklog\.G or bklog\.L instead of logrus directly)?$'
+      - '^context\.WithCancel(# use context\.WithCancelCause instead)?$'
```

Comment:
This is great!  Should we require `context.WithTimeoutCause` and `context.WithDeadlineCause` in this PR also?  (I don't think we use WithDeadline anywhere, but should probably forbid it for the future)

Reply:
yes, marked in TODO comment

Reply:
Actually, I'm not sure if anything can be done for this. Looks like `WithTimeoutCause/WithDeadlineCause` have a completely different signature and cause is added as a parameter instead of returning `func(error)`. This isn't really useful as it is basically just doing `ctx.WithValue("cause", cause)` and doesn't detect when the cancellation actually happened.

Note that if context actually reached timeout then this is somewhat expected as it will happen sometimes in stdlib but if `CancelFunc` gets called then with this stdlib function it does not allow detecting the location.

Maybe for the `defer()` case still something can be done about this.

Reply:
I guess maybe something like this could work?

```
cause := &timeoutCause{}
ctx, cancel := context.WithTimeoutCause(ctx, ..., cause.Init())
cancelErr = cause.WithCancel(cancel)


cancelErr(errors.WithStack(context.DeadlineExceeded))


type timeoutCause {
  error
}

func (t *timeoutCause) Init() {
  // todo: atomic
  // store the context creation stack. this will be shown when timeout is reached and runtime cancels
  t.error = errors.WithStack(context.DeadlineExceeded)
}

func (t *timeoutCause) WithCancel(c func()) func(error) {
  return func(e error) {
    // todo: atomic
    // store the context cancellation stack. this will be shown when cancel is called manually
    t.error = errors.WithStack(o)
  }
}
```

Reply:
In the go issue https://github.com/golang/go/issues/56661  there is also another workaround mentioned.

```
ctx, cancel := context.WithCancelCause(context.Background())
ctx, _ = context.WithTimeoutCause(ctx, 1time.Second, tooSlow)
```

what is simpler but I'm afraid this will cause a linter error for leaking context that needs to be disabled all the time. Really don't understand why `WithTimeoutCause` just does not return `CancelCauseFunc`.

Reply:
Hmm for `WithTimeoutCause` I use something like:

Reply:
This will give you the stacktrace to place that created the context. But for example if you have:
```
{
ctx, cancel := context.WithTimeoutCause(ctx, time.Minute*5, errors.WithStack(context.DeadlineExceeded))
defer cancel()

go func() {
  // leaky goroutine
  err := foo(ctx)
  // cancellation error because context was discarded but no stacktrace to where it happened
}()

return nil
}
```

Reply:
I see, yeah, maybe add a helper function to wrap timeouts, and add that to the forbidigo rule?
```
func withTimeout(ctx context.Context, d time.Duration) (context.Context, func(error)) {
    ctx, cancel := context.WithCancelCause(ctx)
    ctx, _ = context.WithTimeoutCause(ctx, d, errors.WithStack(context.DeadlineExceeded))
    return ctx, cancel
}
```

Reply:
I turns out the linter only has this leak detection for `context.WithTimeout` but nobody has yet added a rule for `context.WithTimeoutCause` yet. So this is a problem for someone in the future trying to update the linter 😉